components/ai_chat/ai_chat_tab_helper.cc

```diff
@@ -308,55 +326,84 @@ void AIChatTabHelper::MakeAPIRequestWithConversationHistoryUpdate(
 
-  ai_chat_api_->QueryPrompt(
-      base::BindOnce(&AIChatTabHelper::OnAPIResponse, base::Unretained(this),
-                     is_summarize_prompt),
-      std::move(prompt_with_history));
+  auto data_received_callback = base::BindRepeating(
+      &AIChatTabHelper::OnAPIStreamDataReceived, base::Unretained(this));
+
+  auto data_completed_callback =
+      base::BindOnce(&AIChatTabHelper::OnAPIStreamDataComplete,
+                     base::Unretained(this), is_summarize_prompt);
+
+  is_request_in_progress_ = true;
+  ai_chat_api_->QueryPrompt(std::move(data_received_callback),
+                            std::move(data_completed_callback),
+                            std::move(prompt_with_history));
 }
 
-void AIChatTabHelper::OnAPIResponse(bool contains_summary,
-                                    const std::string& assistant_input,
-                                    bool success) {
-  SetRequestInProgress(false);
+bool AIChatTabHelper::IsRequestInProgress() {
+  DCHECK(ai_chat_api_);
 
-  if (!success) {
-    // TODO(petemill): show error state separate from assistant message
-    AddToConversationHistory(ConversationTurn{
-        CharacterType::ASSISTANT, ConversationTurnVisibility::VISIBLE,
-        l10n_util::GetStringUTF8(IDS_CHAT_UI_API_ERROR)});
+  return is_request_in_progress_;
+}
 
+void AIChatTabHelper::OnAPIStreamDataReceived(
+    data_decoder::DataDecoder::ValueOrError result) {
+  if (!result.has_value() || !result->is_dict()) {
     return;
   }
 
-  ConversationTurn turn = {CharacterType::ASSISTANT,
-                           ConversationTurnVisibility::VISIBLE,
-                           assistant_input};
+  if (const std::string* completion = result->FindStringKey("completion")) {
+    UpdateOrCreateLastAssistantEntry(*completion);
 
-  if (contains_summary && !assistant_input.empty()) {
-    SetArticleSummaryString(assistant_input);
+    // Trigger an observer update to refresh the UI.
+    for (auto& obs : observers_) {
+      obs.OnAPIRequestInProgress(IsRequestInProgress());
+    }
   }
-
-  AddToConversationHistory(turn);
 }
 
-void AIChatTabHelper::SetRequestInProgress(bool in_progress) {
-  is_request_in_progress_ = in_progress;
+void AIChatTabHelper::OnAPIStreamDataComplete(
+    bool is_summarize_prompt,
+    api_request_helper::APIRequestResult result,
+    bool success) {
+  if (is_summarize_prompt && success && !chat_history_.empty()) {
+    const ConversationTurn& last_turn = chat_history_.back();
+    if (last_turn.character_type == CharacterType::ASSISTANT) {
+      SetArticleSummaryString(last_turn.text);
+    }
+  }
+
+  if (!success || !result.Is2XXResponseCode()) {
+    // TODO(petemill): show error state separate from assistant message
+    AddToConversationHistory(ConversationTurn{
+        CharacterType::ASSISTANT, ConversationTurnVisibility::VISIBLE,
+        l10n_util::GetStringUTF8(IDS_CHAT_UI_API_ERROR)});
+  }
+
+  is_request_in_progress_ = !success;
```

Comment:
I'm still not following. In which case do we want the API request to complete but still want is_request_in_progress to stay `true`?

Reply:
In case of a network failure or error, it is possible for the API to technically complete but overall request may still considered as "in progress" to the user. for ex, in the middle of text generation showing a blinking caret